### PR TITLE
Support API Gateway

### LINKED ISSUE #1077: Support API Gateway
#### Description

Add Cortex API configuration to configure AWS API Gateway, so that users can have out-of-the-box access to trusted SSL certificates.

### PR DESCRIPTION
**Pull request by tthebst**
closes #1077

---

checklist:

- [x] run `make test` and `make lint`
- [x] test manually (i.e. build/push all images, restart operator, and re-deploy APIs)
- [ ] update examples
- [ ] update docs and add any new files to `summary.md` (view in [gitbook](https://docs.cortex.dev/v/master) after merging)
- [x] cherry-pick into release branches if applicable
- [x] alert the dev team if the dev environment changed

### CODE REVIEW

File: manager/create_gateway_integration.py [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+# limitations under the License.
+
+import boto3
+import sys
+import traceback
+import os
+
+
+def get_istio_api_gateway_elb_arn():
+    for elb in client_elb.describe_load_balancers()["LoadBalancers"]:
```
[the remark this answers is not in the document]
Comment (tthebst) [reply manager/create_gateway_integration.py:22]: good idea

File: pkg/lib/aws/apigateway.go [hunk 2]
```diff
@@ -0,0 +1,359 @@
...
+	"fmt"
+	"strings"
+
+	"github.com/aws/aws-sdk-go/service/apigatewayv2"
+	"github.com/cortexlabs/cortex/pkg/lib/errors"
+)
+
+// GetVpcLinkID vpc-link id only exists if internal facing loadbalancer
+func (c *Client) getVpcLinkID(clusterName string) (string, error) {
+	vpcLinks, err := c.APIGatewayv2().GetVpcLinks(&apigatewayv2.GetVpcLinksInput{})
```
[the remark this answers is not in the document]
Comment (tthebst) [reply pkg/lib/aws/apigateway.go]: The maximum number of returned results per page. The default value is 25
and the maximum value is 500. 
Do you think it is enough if we set it to 500?

[the remark this answers is not in the document]
Comment (deliahu) [reply pkg/lib/aws/apigateway.go]: Yes that would be safest

File: docs/cluster-management/config.md [hunk 3]
```diff
@@ -73,7 +76,7 @@ tags:  # <string>: <string> map of key/value pairs
 # see https://docs.cortex.dev/v/master/cluster-management/spot-instances for additional details on spot configuration
 spot: false
 
-# see https://docs.cortex.dev/v/master/guides/subdomain-https-setup for instructions on how to set up HTTPS for APIs
+# see https://docs.cortex.dev/v/master/guides/custom-domain for instructions on how to set up a custom domain
```
Comment (tthebst) [inline docs/cluster-management/config.md:79]: I think we can simplify this by making a guide on how to add custom [domains](https://docs.aws.amazon.com/apigateway/latest/developerguide/http-api-custom-domain-names.html) with API gateway which is easier!

File: docs/cluster-management/uninstall.md [hunk 4]
```diff
@@ -43,6 +43,4 @@ aws s3 rb --force s3://<bucket>
 aws logs describe-log-groups --log-group-name-prefix=<log_group> --query logGroups[*].[logGroupName] --output text | xargs -I {} aws logs delete-log-group --log-group-name {}
 ```
 
-If you've setup API gateway and want to delete it, please follow these [instructions](../guides/api-gateway.md#cleanup).
-
-If you've configured HTTPS by specifying an SSL Certificate for a subdomain in your cluster configuration, you may wish to remove the SSL Certificate and Hosted Zone for the domain by following these [instructions](../guides/subdomain-https-setup.md#cleanup).
+If you've configured a custom domain for your APIs, you may wish to remove the SSL Certificate and Hosted Zone for the domain by following these [instructions](../guides/custom-domain.md#cleanup).
```
Comment (tthebst) [inline docs/cluster-management/uninstall.md:46]: I think we can simplify this by making a guide on how to add custom [domains](https://docs.aws.amazon.com/apigateway/latest/developerguide/http-api-custom-domain-names.html) with API gateway which is easier!

File: docs/deployments/networking.md [hunk 5]
```diff
@@ -0,0 +1,11 @@
...
+
+## API Gateway
+
+APIs are deployed with an internet-facing API Gateway by default (the API Gateway forwards requests to the API load balancer). Each API can be independently configured to not create the API Gateway endpoint by setting `api_gateway: none` in the `networking` field of the [api configuration](api-configuration.md). If the API Gateway endpoint is not created, your API can still be accessed via the API load balancer; `cortex get API_NAME` will show the load balancer endpoint if API Gateway is disabled.
+
+If you'd like to force traffic to go through your public API Gateway endpoint, set `api_load_balancer_scheme: internal` in your [cluster configuration](../cluster-management/config.md) file before creating your cluster.
+
+If you'd like to ensure that an API is not publicly accessible, set `api_load_balancer_scheme: internal` in your [cluster configuration](../cluster-management/config.md) file, and set `api_gateway: none` in the `networking` field of your [api configuration](api-configuration.md). If you do this, you will need to configure [VPC Peering](../guides/vpc-peering.md) to make prediction requests to your APIs.
```
Comment (tthebst) [inline docs/deployments/networking.md:11]: maybe a table?
                      HTTPS ENDPOINT                                  | HTTP ENDPOINT
____________|______________________________________________________ ___ 
API Public   | api_load_balancer_scheme: internal`   | api_load_balancer_scheme: internet facing`
                    | api_gateway: public                               |   api_gateway: none
____________ |______________________________________________________________
API Private  | ------------------------------------   | api_load_balancer_scheme: internal`
                    | ------------------------------- ----- |   api_gateway: none

Comment (deliahu) [reply docs/deployments/networking.md:11]: Good idea, I just added it!

File: pkg/lib/errors/error.go [hunk 6]
```diff
@@ -0,0 +1,196 @@
+/*
```
Comment (deliahu) [inline pkg/lib/errors/error.go:1]: note: this file did not change, it was just renamed from `errors.go`

**Pull request merged**